diff --git a/third_party/SimpleAmqpClient-2.4.0/CMakeLists.txt b/third_party/SimpleAmqpClient-2.4.0/CMakeLists.txt
index d8d33c1..95aaa58 100644
--- a/third_party/SimpleAmqpClient-2.4.0/CMakeLists.txt
+++ b/third_party/SimpleAmqpClient-2.4.0/CMakeLists.txt
@@ -32,7 +32,9 @@ string(REGEX MATCH "[0-9]+" _API_VERSION_PATCH ${_API_VERSION_PATCH})
 
 set(SAC_APIVERSION ${_API_VERSION_MAJOR}.${_API_VERSION_MINOR}.${_API_VERSION_PATCH})
 
-FIND_PACKAGE(Boost 1.47.0 COMPONENTS chrono system REQUIRED)
+set(BOOST_LIBRARYDIR /root/contrail/build/lib)
+set(BOOST_INCLUDEDIR /root/contrail/build/include)
+FIND_PACKAGE(Boost 1.69.0 COMPONENTS chrono system REQUIRED)
 INCLUDE_DIRECTORIES(SYSTEM ${Boost_INCLUDE_DIRS})
 
 SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/Modules)
