diff --git a/third_party/log4cplus-2.1.1/src/global-init.cxx b/third_party/log4cplus-2.1.1/src/global-init.cxx
index c27ad7b..1beb0bb 100644
--- a/third_party/log4cplus-2.1.1/src/global-init.cxx
+++ b/third_party/log4cplus-2.1.1/src/global-init.cxx
@@ -153,14 +153,18 @@ instantiate_thread_pool ()
 //! therefore the ThreadPool can only be destroyed after that.
 struct ThreadPoolHolder
 {
+#if !defined(LOG4CPLUS_SINGLE_THREADED)
     std::atomic<progschj::ThreadPool*> thread_pool{};
+#endif
 
     ThreadPoolHolder () = default;
     ThreadPoolHolder (ThreadPoolHolder const&) = delete;
     ~ThreadPoolHolder ()
     {
+#if !defined(LOG4CPLUS_SINGLE_THREADED)
         auto const tp = thread_pool.exchange(nullptr, std::memory_order_release);
         delete tp;
+#endif
     }
 };
 
